File: backend/utilities/check_coverage.py
import json
import os
import sys

# Default threshold for files under app/
DEFAULT_THRESHOLD = 70.0

# Specific file thresholds (e.g. security helpers)
SPECIFIC_THRESHOLDS = {
    "app/core/crypto_helper.py": 80.0,
}

# Exemptions / Allowed lower coverage due to infrastructure mocking
EXEMPTIONS = {
    "app/services/docker_manager.py": 60.0,
    "app/services/huggingface_downloader.py": 40.0,
}

def main():
    coverage_file = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "coverage.json"))
    if not os.path.exists(coverage_file):
        print(f"Error: Coverage JSON file not found at {coverage_file}", file=sys.stderr)
        print("Please run pytest with coverage first (e.g. pytest --cov=app --cov-report=json)", file=sys.stderr)
        sys.exit(1)

    with open(coverage_file, "r") as f:
        data = json.load(f)

    files_data = data.get("files", {})
    failed = False
    checked_count = 0

    print("Checking per-file code coverage...")
    print("=" * 60)

    for file_path, file_info in sorted(files_data.items()):
        # Only check files in the app/ directory
        if not file_path.startswith("app/"):
            continue

        # Ignore empty or __init__.py files with 0 statements
        summary = file_info.get("summary", {})
        if summary.get("num_statements", 0) == 0:
            continue

        percent_covered = summary.get("percent_covered", 0.0)
        
        # Determine expected threshold
        if file_path in SPECIFIC_THRESHOLDS:
            expected = SPECIFIC_THRESHOLDS[file_path]
            desc = " (Custom minimum)"
        elif file_path in EXEMPTIONS:
            expected = EXEMPTIONS[file_path]
            desc = " (Exemption)"
        else:
            expected = DEFAULT_THRESHOLD
            desc = ""

        checked_count += 1

        if percent_covered < expected:
            print(f"FAIL: {file_path}")
            print(f"      Coverage: {percent_covered:.2f}% < Expected: {expected:.2f}%{desc}")
            failed = True
        else:
            print(f"PASS: {file_path} ({percent_covered:.2f}% >= {expected:.2f}%{desc})")

    print("=" * 60)
    if failed:
        print("Coverage check FAILED! Some files do not meet their coverage threshold.", file=sys.stderr)
        sys.exit(1)
    else:
        print(f"Coverage check PASSED! All {checked_count} checked files met their thresholds.")
        sys.exit(0)

if __name__ == "__main__":
    main()

File: frontend/utilities/check_coverage.js
const fs = require('fs');
const path = require('path');

// Default threshold for all metrics (statements, branches, functions, lines)
const DEFAULT_THRESHOLD = 70.0;

// Specific file thresholds (if any need to be higher, e.g. critical modules)
const SPECIFIC_THRESHOLDS = {
  // e.g. 'src/app/page.tsx': 80.0
};

// Exemptions / Allowed lower thresholds
const EXEMPTIONS = {
  // No files currently need exemptions below 70% as all exceed it!
};

function main() {
  const summaryFile = path.resolve(__dirname, '..', 'coverage', 'coverage-summary.json');
  if (!fs.existsSync(summaryFile)) {
    console.error(`Error: Coverage summary JSON file not found at ${summaryFile}`);
    console.error('Please run npm run test:coverage first.');
    process.exit(1);
  }

  const data = JSON.parse(fs.readFileSync(summaryFile, 'utf8'));
  let failed = false;
  let checkedCount = 0;

  console.log('Checking frontend per-file code coverage...');
  console.log('='.repeat(60));

  const frontendDir = path.resolve(__dirname, '..');

  for (const [absPath, fileInfo] of Object.entries(data)) {
    if (absPath === 'total') continue;

    // Convert absolute path to project-relative path (e.g. src/components/Sidebar.tsx)
    const relPath = path.relative(frontendDir, absPath);

    // Only check files in the src/ directory
    if (!relPath.startsWith('src/')) {
      continue;
    }

    const { lines, statements, functions, branches } = fileInfo;

    // Determine expected threshold
    let expected = DEFAULT_THRESHOLD;
    let desc = '';
    if (relPath in SPECIFIC_THRESHOLDS) {
      expected = SPECIFIC_THRESHOLDS[relPath];
      desc = ' (Custom minimum)';
    } else if (relPath in EXEMPTIONS) {
      expected = EXEMPTIONS[relPath];
      desc = ' (Exemption)';
    }

    const linePct = lines.pct;
    const stmtPct = statements.pct;
    const funcPct = functions.pct;
    const branchPct = branches.pct;

    checkedCount++;

    if (linePct < expected || stmtPct < expected || funcPct < expected || branchPct < expected) {
      console.error(`FAIL: ${relPath}`);
      console.error(`      Lines: ${linePct.toFixed(2)}%, Statements: ${stmtPct.toFixed(2)}%, Functions: ${funcPct.toFixed(2)}%, Branches: ${branchPct.toFixed(2)}%`);
      console.error(`      Expected: >= ${expected.toFixed(2)}%${desc}`);
      failed = true;
    } else {
      console.log(`PASS: ${relPath}`);
      console.log(`      (Lines: ${linePct.toFixed(2)}%, Branches: ${branchPct.toFixed(2)}% >= ${expected.toFixed(2)}%${desc})`);
    }
  }

  console.log('='.repeat(60));
  if (failed) {
    console.error('Coverage check FAILED! Some files do not meet their coverage threshold.');
    process.exit(1);
  } else {
    console.log(`Coverage check PASSED! All ${checkedCount} checked files met their thresholds.`);
    process.exit(0);
  }
}

main();
